agent/src/agent_win.cc

```diff
@@ -11,87 +11,383 @@
+  DWORD err = ERROR_SUCCESS;
+  DWORD count;
+  BOOL success = GetOverlappedResult(handle, &overlapped_, &count, FALSE);
+  if (!is_connected_) {
+    if (success) {
+      // A Google Chrome browser connected to the agent.
+      is_connected_ = true;
+      buffer_.resize(kBufferSize);
+
+      err = BuildBrowserInfo();
+      if (err == ERROR_SUCCESS) {
+        handler_->OnBrowserConnected(browser_info_);
+      }
+    } else {
+      err = GetLastError();
+    }
+  } else {
+    // A new message has arrived from Google Chrome.
+    err = OnReadFile(success, count);
+  }
+
+  // If all data has been read, queue another read.
+  if (err == ERROR_SUCCESS || err == ERROR_MORE_DATA) {
+    err = QueueReadFile(err == ERROR_SUCCESS);
+  }
+
+  if (err != ERROR_SUCCESS && err != ERROR_IO_PENDING &&
+      err != ERROR_MORE_DATA) {
+    Cleanup();
+  } else {
+    // Don't propagate all the "success" error codes to the called to keep
+    // this simpler.
+    err = ERROR_SUCCESS;
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::CreatePipe(
+    const std::string& name,
+    bool is_first_pipe,
+    HANDLE* handle) {
+  DWORD err = ERROR_SUCCESS;
+
+  DWORD mode = PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED;
+  if (is_first_pipe) {
+    mode |= FILE_FLAG_FIRST_PIPE_INSTANCE;
+  }
+  *handle = CreateNamedPipeA(name.c_str(),
+    mode,
+    PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT |
+    PIPE_REJECT_REMOTE_CLIENTS, PIPE_UNLIMITED_INSTANCES, kBufferSize,
+    kBufferSize, 0, nullptr);
+  if (*handle == INVALID_HANDLE_VALUE) {
+    err = GetLastError();
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::ConnectPipe() {
+  // In overlapped mode, connecting to a named pipe always returns false.
+  if (ConnectNamedPipe(handle_, &overlapped_)) {
+    return GetLastError();
+  }
+
+  DWORD err = GetLastError();
+  if (err == ERROR_IO_PENDING) {
+    // Waiting for a Google Chrome Browser to connect.
+    err = ERROR_SUCCESS;
+  } else if (err == ERROR_PIPE_CONNECTED) {
+    // A Google Chrome browser is already connected.  Make sure event is in
+    // signaled state in order to process the connection.
+    if (SetEvent(overlapped_.hEvent)) {
+      err = ERROR_SUCCESS;
+    } else {
+      err = GetLastError();
+    }
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::ResetInternal(std::string& pipename,
+                                          bool is_first_pipe) {
+  DWORD err = ERROR_SUCCESS;
+
+  // If this is the not the first time, disconnect from any existing Google
+  // Chrome browser.  Otherwise creater a new pipe.
+  if (handle_ != INVALID_HANDLE_VALUE) {
+    if (!DisconnectNamedPipe(handle_)) {
+      err = GetLastError();
+    }
+  } else {
+    err = CreatePipe(pipename, is_first_pipe, &handle_);
+  }
+
+  // Make sure event starts in reset state.
+  if (err == ERROR_SUCCESS && !ResetEvent(overlapped_.hEvent)) {
+    err = GetLastError();
+  }
+
+  if (err == ERROR_SUCCESS) {
+    err = ConnectPipe();
+  }
```

Comment:
Double check: Is it expected that both `err = GetLastError();` and `err = ConnectPipe();`be called if `err == ERROR_SUCCESS && !ResetEvent(overlapped_.hEvent)` are met?